README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# place

<!-- badges: start -->

[![R-CMD-check](https://github.com/mdsumner/place/workflows/R-CMD-check/badge.svg)](https://github.com/mdsumner/place/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/place)](https://CRAN.R-project.org/package=place)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of place is to specify a region simply as point
(longitude,latitude) and a region around that.

This package does absolutely nothing with this information, but other
packages might (get where I’m going …?).

Data at [virtualearth](http://a0.ortho.tiles.virtualearth.net/tiles/).

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("mdsumner/place")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(place)



## get me outta here
where()
#> where : 88.63528, 27.45667 (lon, lat)
#> width : 111120
#> height: 111120
```

Ok, get me somewhere specific.

``` r
where(156, 34, wh = 100000)  ## 100km either side of that location
#> where : 156, 34 (lon, lat)
#> width : 1e+05
#> height: 1e+05
```

More useful, give me a defined discretized region around a location.

``` r
where(147, -42, wh = c(1, 1) * 5000)
#> where : 147, -42 (lon, lat)
#> width : 5000
#> height: 5000

place("1 George St, Bathurst NSW", wh = c(1, 1.6) * 15000, dimension = 1024)
#> Loading required namespace: tidygeocoder
#> $extent
#> [1] -15000  15000 -24000  24000
#> 
#> $dimension
#> [1] 1024 1638
#> 
#> $projection
#> [1] "+proj=laea +lon_0=149.6 +lat_0=-33.4 +datum=WGS84"
```

------------------------------------------------------------------------

## Code of Conduct

Please note that the place project is released with a [Contributor Code
of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# place 0.0.0.9002

* Allow input of actual projection string to 'family'.

* New data set `city`. 

* New function `place()`. 

* Rename. 

# where 0.0.0.9000

* New function `extent_grid()` to generate sensible aspect ratio for an extent. 


R/where.R
#' Where
#'
#' Specify a region by location point (longitude, latitude) and a width and height
#' around that point.
#'
#' Width and height will be assumed to be in metres usually, but no units are recorded.
#'
#' If no input point is given one is generated randomly. The default wh value is approximately the distance
#' along one degree of a great circle on Earth.
#' @param x longitude
#' @param y latitude
#' @param wh width and height (can be a single value) assumed to be metres
#' @param ... in the print method
#'
#' @return a where structure, the location and width/height around it
#' @export
#' @importFrom grDevices xy.coords
#' @importFrom stats runif
#' @examples
#' where()
#'
#' where(147, -42, wh = c(2000, 3000))
where <- function(x, y = NULL, wh = 1.852e3 * 60) {
  if (missing(x) && is.null(y)) {
    x <- runif(1L, -180, 180)
    y <- runif(1L, -90, 90)
  }
  if (is.null(y)) {
    if (length(x) < 2) stop("both x, and y (longitude,latitude) must be given")
  }

  if (length(x) == 2) {
    xy <- x
  } else {
    xy <- do.call(cbind, xy.coords(x, y)[c("x", "y")])[1L, , drop = FALSE]  ## we only get one
  }
  wh <- rep(wh, length.out = 2L)
  ## should we do this??  NO
  #if (length(wh) < 2) wh <- c(1/cos(xy[2L] * pi/180), 1) * wh
  if (abs(xy[1L])> 180) message("where x is outside the normal longitude range -180,180")
  if (abs(xy[2L])> 90) message("where y is outside the normal latitude range -90,90")
  if (anyNA(xy)) stop("missing values x,y")
  if (anyNA(wh[1])) stop("missing value wh")
  structure(list(where = xy, wh = wh), class = c("where", "list"))
}

## we want +, - and 1-way or two way to increase the width height
## possibly *, / to shift sideways, up-down?
Ops.where <- function(e1, e2) {
 stop("not implemented")
}
#' @name where
#' @export
print.where <- function(x, ...) {
  cat(sprintf("where : %s, %s (lon, lat)\n", format(x$where[1L]), format(x$where[2L])))
  cat(sprintf("width : %s\n", format(x$wh[1L])))
  cat(sprintf("height: %s\n", format(x$wh[2L])))
}
#' @name dim
#' Export
dim.where <- function(x) {
  c(1L, 1L)
}


R/extent_grid.R
#' Extent grid
#'
#' Helper function to create a grid dimension with a sensible aspect ratio.
#'
#' @param extent numeric vector, xmin,xmax,ymin,ymax
#' @param dim optional dimension, 1 number or 2
#' @param projection optional projection string to store
#' @param lonlat if set uses cosine latitude scaling for width for longitude
#'
#' @return list of extent, dimension, projection
#' @export
#' @importFrom grDevices dev.cur dev.size
#' @examples
#' extent_grid(c(0, 1, 0, 2))
extent_grid <- function(extent, dim = NULL, projection = NULL, lonlat = FALSE) {
  xl <- diff(extent[1:2])
  yl <- diff(extent[3:4])
  asp <- xl/yl * if (lonlat) cos(mean(yl) * pi/180) else 1
  if (is.null(dim) && !dev.cur() == 1) {
    dim <- dev.size("px")
  } else {
    if (is.null(dim)) dim <- c(512, 512)
  }

  # dim <- rep(dim, length.out = 2L)
  # wm <- which.max(dim)
  # dim[-wm] <- max(dim)*asp

  dim <- as.integer(dim * c(1, 1) * sort(c(1, asp)))

  list(extent = extent, dimension = as.integer(round(dim)), projection = projection)
}


R/place.R

#' Place
#'
#' Family
#'
#' Note that if you give a longlat family, you should give a wh to where in degrees (we might fix)
#'
#' @param x place name (a city)
#' @param dimension number of pixels per margin (length one and aspect ratio is automatic)
#' @param family projection family 'laea' default, or an actual projection string
#' @param ... other arguments passed to [where]
#'
#'
#' @return list of extent, dimension, projection
#' @export
#'
#' @examples
#' place("Hobart")
#' place("Bathurst", wh = c(10000, 15000))
#' if(requireNamespace("tidygeocoder")) place("Downtown Juneau, Juneau, AK, USA")
place <- function(x, dimension = 512L, family = "laea", ...) {
 UseMethod("place")
}
#' @name place
#' @export
place.character <- function(x, dimension = 512L, family = "laea", ...) {

  ## first, is it in our city/s
  idx <- match(tolower(x[1L]), tolower(city$name))
  if (is.na(idx) || length(idx) < 1 || idx < 1) {
    if (requireNamespace("tidygeocoder")) {
      data <- try(tidygeocoder::geo(x, quiet = TRUE, progress_bar = FALSE), silent = TRUE)
      if (inherits(data, "try-error") || is.na(data$lat[1L])) {
        #message()
        stop(sprintf("could not geocode %s", x))
      }
      w <- where(data$long[1], data$lat[1], ...)
    } else {
      message(sprintf("could not attempt to geocode %s, install 'tidygeocoder' and try again", x))
      stop(sprintf("cannot find %s", x))
    }
  } else {
    w <- where(city$lon[idx], city$lat[idx], ...)
  }
  ##dput(trimws( unlist(lapply(strsplit(system("proj -lp | grep conic --ignore-case", intern = T), ":"), "[", 1L))))
  conics <- c("bipc", "ccon", "eqdc", "imw_p", "lcc", "lcca", "leac", "pconic", "poly", "rpoly")

  if (grepl(".*\\+.*", family) || grepl("[0-9]", family) ) {
    ## assume we gave an actual projection string +proj, AUTH:, or wkt
    crs <- family
    if (grepl("3857", family ) || grepl("merc", family)) {
    warning("place heuristics for the projection don't work with Mercator yet")

    }
  } else {
      if (family %in% c("merc", "eqc")) {
    warning("place heuristics for the projection don't work with Mercator or Equidistant Cylindrical (yet)")
  }

  if (family %in% conics) {
    lat_1_2 <- w$where[2] + c(-1, 1) * w$wh[2] * cos(w$where[2] * pi/180) / 111120
    crs <- sprintf("+proj=%s +lon_0=%f +lat_0=%f +lat_1=%f +lat_2=%f +datum=WGS84", family, w$where[1], w$where[2], lat_1_2[1], lat_1_2[2])
  } else {
    crs <- sprintf("+proj=%s +lon_0=%f +lat_0=%f +datum=WGS84", family, w$where[1], w$where[2])
  }


  }
  ## here do aspect ratio, because we might have a where()$wh aspect ratio
  if (length(dimension) == 1L) {
    rat <- w$wh[2]/w$wh[1]
    dimension <- round(rep(dimension, 2) * sort(c(1, rat)))
  }
  list(extent = c(-1, 1, -1, 1) * rep(w$wh, each = 2), dimension = dimension, projection = crs, point = w$where )
}
